Controller álbuns e músicas

ThallysRocha opened the pull request:
# Controller álbuns e músicas
- Foram criadas as rotas, do álbum e músicas

ThallysRocha committed:
alterações no model para estabelecer relações

ThallysRocha committed:
relações estabelecidas

ThallysRocha committed:
feat: check if featuring artist is the song owner

SamuelSimoes31 commented on back/src/controllers/albums.js:
```diff
@@ -1,14 +1,25 @@
 
 albumsRouter.post('',async (request, response) => {
-    const {name,image} = request.body;
+    const {name,image,year,createdAt,songs} = request.body;
```
O createdAt é gerado automaticamente, não tem que ser passado na requisição.

SamuelSimoes31 commented on back/src/controllers/albums.js:
```diff
@@ -1,14 +1,25 @@
-    const {name,image} = request.body;
+    const {name,image,year,createdAt,songs} = request.body;
     try {   
         if(await Album.findOne({ name })) return response.status(400).send({ error: 'Album already exists' });
```
Dois artistas diferentes não podem ter o mesmo nome de álbum?

ThallysRocha replied:
Pensando bem, vc está certo. Só não poderia o mesmo artista ter dois álbuns com o mesmo nome.

SamuelSimoes31 commented on back/src/controllers/albums.js:
```diff
@@ -1,14 +1,25 @@
+
         return response.send({ album });
     } catch (error) {
         return response.status(400).send({ error: 'Register failed' });
```
Só essa mensagem de erro não tá sendo suficiente para dizer qual foi o erro. Aí ou adiciona mais retornos ao longo da função, ou imprime o error do catch.

SamuelSimoes31 commented on back/src/controllers/songs.js:
```diff
@@ -7,15 +7,16 @@ export const songsRouter = express.Router();
             if(await Song.findOne({ url })) return response.status(400).send({ error: 'Song already exists' });
-            if(!participations||await Artist.findOne({ 'name' : participations })){
+            const featuringArtist = await Artist.findOne({ 'name' : participations })
+            if(!participations||(featuringArtist._id != request.userId)){
```
!==

ThallysRocha replied:
!=

ThallysRocha replied:
Um id é string, e o outro não.

SamuelSimoes31 commented on back/src/controllers/albums.js:
```diff
@@ -1,14 +1,25 @@
-        const album = await Album.create(request.body);
+        //if(image&&!/\.(jpg|jpeg|png|webp|avif|gif|svg)$/.test(image)) return response.status(400).send({ error: 'Image format is not valid' });
+        const album = await Album.create({name, image, year, createdAt, artist : request.userId});
+        await Promise.all(songs.map(async song =>{
```
Se songs não por passado o songs.map dá pau. Mas a o álbum aparetement é criado.

SamuelSimoes31 commented on back/src/controllers/albums.js:
```diff
@@ -1,14 +1,25 @@
+        //if(image&&!/\.(jpg|jpeg|png|webp|avif|gif|svg)$/.test(image)) return response.status(400).send({ error: 'Image format is not valid' });
+        const album = await Album.create({name, image, year, createdAt, artist : request.userId});
+        await Promise.all(songs.map(async song =>{
+            const albumSong = new Song({...song, album: album._id}); 
```
delete song.createdAt só para garantir que esse campo n é passado na criação. Tem q ser gerado automaticamente.

SamuelSimoes31 commented on back/src/models/song.js:
```diff
@@ -16,6 +16,16 @@ const SongSchema = new mongoose.Schema({
+  artist: {
+    type : mongoose.Schema.Types.ObjectId,
+    ref : 'Artist',
+    require : true,
```
Aqui é require ou required?

SamuelSimoes31 commented on back/src/controllers/songs.js:
```diff
@@ -7,15 +7,16 @@ export const songsRouter = express.Router();
 songsRouter.post('',async (request, response) => {
```
A rota tá meio que totalmente desaçociada do album e do artista. Tem que rever o momento em que cada rota vai acontecer.

ThallysRocha committed:
Fix: checar se o álbum já foi cadastrado

Foi retirado o campo de createdAt.

ThallysRocha committed:
Fix: validation model

ThallysRocha committed:
feat: checar se algum campo de música está vazio

ThallysRocha committed:
Feat: passar um álbum válido ao criar uma música

Para criar uma música independentemente da criação do álbum é possível, se for passado o id de um álbum que está contido na lista de álbuns do artista logado. Além de ajustar a mensagem do catch.

ThallysRocha committed:
Fix: validation model

Ajustando o required no model de músicas.

ThallysRocha committed:
Fix: mensagem do catch
